File: combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        resli = []
        
        def dfs(candidates, res, target):
            for x, i in enumerate(candidates):
                if i > target:
                    break
                elif i == target:
                    resli.append(res)
                    resli[-1].append(i)
                else:
                    dfs(candidates[x:] ,res+[i], target-i)
        
        dfs(candidates, [], target)

        return resli
    

File: combination-sum/combination-sum.java
class Solution {
    int[] candidates;
    List<List<Integer>> ret = new ArrayList<List<Integer>>();
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        this.candidates = candidates;
        backtrack(0, target, new ArrayList<Integer>(), 0);
        return this.ret;
    }
    
    public void backtrack(int idx, int target, List<Integer> curr, int sum) {
        if(idx == candidates.length) return;
        if(sum > target) return;
        if(sum == target) {
            ret.add(new ArrayList<Integer>(curr));
        }
        for(int i=idx; i<candidates.length; i++) {
            // choose
            curr.add(candidates[i]);
            backtrack(i, target, curr, sum+candidates[i]);
            curr.remove(curr.size()-1);
        }
        return;
    }
    
    
} 